trajoptlib/src/mecanum_trajectory_generator.cpp

```diff
@@ -0,0 +1,449 @@
+
+  for (size_t index = 0; index < samp_tot; ++index) {
+    Rotation2v<double> θ_k{cosθ.at(index), sinθ.at(index)};
+    Translation2v<double> v_k{vx.at(index), vy.at(index)};
+
+    std::vector<slp::Variable<double>> Fx_body;
+    Fx_body.reserve(path.drivetrain.wheels.size());
+    std::vector<slp::Variable<double>> Fy_body;
+    Fy_body.reserve(path.drivetrain.wheels.size());
+
+    for (size_t i = 0; i < path.drivetrain.wheels.size(); ++i) {
+      Fx_body.push_back(fx_coeff[i] * F.at(index)[i]);
+      Fy_body.push_back(fy_coeff[i] * F.at(index)[i]);
+    }
+
+    // sum forces to get net body force
+    auto Fx_net = std::accumulate(Fx_body.begin(), Fx_body.end(), slp::Variable<double>{0.0});
+    auto Fy_net = std::accumulate(Fy_body.begin(), Fy_body.end(), slp::Variable<double>{0.0});
+
+    // solve for net torque
+    slp::Variable<double> τ_net = 0.0;
+    for (size_t module_index = 0; module_index < path.drivetrain.wheels.size();
+         ++module_index) {
+      const auto& wheel_pos = path.drivetrain.wheels.at(module_index);
+      // force vector in body frame
+      Translation2v<double> F_body{Fx_body.at(module_index), Fy_body.at(module_index)};
+
+      // torque = r * F
+      // computed in body frame: r_body * F_body
+      τ_net += wheel_pos.cross(F_body);
+    }
+
+    // power constraint
+    auto v_wrt_robot = v_k.rotate_by(-θ_k);
+    for (size_t module_index = 0; module_index < path.drivetrain.wheels.size();
+         ++module_index) {
+      const auto& wheel_pos = path.drivetrain.wheels.at(module_index);
+
+      double sign = (module_index == 2 || module_index == 3) ? 1.0 : -1.0;
+
+      // v_active_trans = vx * cos(45°) - vy * sin(45°) * sign
+      // = (vx - vy * sign) / sqrt(2)
+      auto v_active_from_translation =
+          (v_wrt_robot.x() - v_wrt_robot.y() * sign) * s;
+
+      //   v_active_rot = v_rot_x*cos(45°) - v_rot_y*sin(45°)*sign
+      //   = (-ω*y)*cos(45°) - (ω*x)*sin(45°)*sign
+      //   = -ω*(y*cos(45°) + x*sin(45°)*sign)
+      //   = -ω*(y + x*sign) / sqrt(2)
+      auto v_active_from_rotation =
+          -ω.at(index) * (wheel_pos.y() + wheel_pos.x() * sign) * s;
+
+      auto v_wheel = v_active_from_translation + v_active_from_rotation;
+
+      double max_wheel_velocity = path.drivetrain.wheel_radius *
+                                  path.drivetrain.wheel_max_angular_velocity;
+
+      // v^2 <= vmax^2
+      problem.subject_to(v_wheel * v_wheel <=
+                         max_wheel_velocity * max_wheel_velocity);
+
+      double max_wheel_force =
+          (path.drivetrain.wheel_max_torque / path.drivetrain.wheel_radius) * s;
+
+      // static force constraint prevents wheel slip
+      double normal_force_per_wheel = path.drivetrain.mass / 4.0 * 9.8;
+      double static_friction_coef = (path.drivetrain.static_friction_coefficient > 0.0)
+          ? path.drivetrain.static_friction_coefficient
+          : path.drivetrain.wheel_cof * 1.2;
```

Comment:
These numbers seem hardcoded. Aren't there already variables in `path.drivetrain` for this stuff?